extensions/android-paging3/src/test/java/com/squareup/sqldelight/android/paging3/KeyedQueryPagingSourceTest.kt

```diff
@@ -0,0 +1,244 @@
+    val source = KeyedQueryPagingSource(
+      queryProvider = this::query,
+      pageBoundariesProvider = this::pageBoundaries,
+      transacter = transacter,
+      dispatcher = TestCoroutineDispatcher(),
+    )
+
+    assertNull(
+      source.getRefreshKey(
+        PagingState(
+          emptyList(),
+          null,
+          PagingConfig(3),
+          0
+        )
+      )
+    )
+  }
+
+  @Test fun `invoking getRefreshKey with loaded first page returns correct result`() {
+    val source = KeyedQueryPagingSource(
+      queryProvider = this::query,
+      pageBoundariesProvider = this::pageBoundaries,
+      transacter = transacter,
+      dispatcher = TestCoroutineDispatcher(),
+    )
+
+    val results = runBlocking { source.load(Refresh(key = null, loadSize = 3, false)) }
+    val refreshKey = source.getRefreshKey(
+      PagingState(
+        listOf(results as LoadResult.Page),
+        null,
+        PagingConfig(3),
+        0
+      )
+    )
+
+    assertEquals(0L, refreshKey)
+  }
+
+  @Test fun `invoking getRefreshKey with single loaded middle page returns correct result`() {
+    val source = KeyedQueryPagingSource(
+      queryProvider = this::query,
+      pageBoundariesProvider = this::pageBoundaries,
+      transacter = transacter,
+      dispatcher = TestCoroutineDispatcher(),
+    )
+
+    val results = runBlocking { source.load(Refresh(key = 6L, loadSize = 3, false)) }
+    val refreshKey = source.getRefreshKey(
+      PagingState(
+        listOf(results as LoadResult.Page),
+        null,
+        PagingConfig(3),
+        0
+      )
+    )
+
+    assertEquals(6L, refreshKey)
+  }
+
+  private fun pageBoundaries(limit: Long): Query<Long> {
+//    val sql = """
+//      |SELECT value
+//      |FROM (
+//      |  SELECT
+//      |    value,
+//      |    CASE
+//      |      WHEN (row_number() OVER(ORDER BY value ASC) - 1) % ? = 0 THEN 1
+//      |      WHEN value = ? THEN 1
+//      |      ELSE 0
+//      |    END page_boundary
+//      |  FROM testTable
+//      |  ORDER BY value ASC
+//      |)
+//      |WHERE page_boundary = 1;
+//    """.trimMargin()
+
+    return object : Query<Long>(mutableListOf(), { cursor -> cursor.getLong(0)!! }) {
+      override fun execute() = object : SqlCursor {
+
+        private val boundaries = (0L until 10L step limit).iterator()
+
+        override fun close() = Unit
+        override fun getBytes(index: Int): ByteArray? = TODO("Not yet implemented")
+        override fun getDouble(index: Int): Double? = TODO("Not yet implemented")
+        override fun getLong(index: Int): Long? = boundaries.next()
+        override fun getString(index: Int): String? = TODO("Not yet implemented")
+        override fun next(): Boolean = boundaries.hasNext()
+      }
+    }
+  }
+
+  private fun query(beginInclusive: Long, endExclusive: Long?): Query<Long> {
+    val sql = """
+      |SELECT value FROM testTable
+      |WHERE value >= ? AND (value < ? OR ? IS NULL)
```

Comment:
you can use indexed params here `:1 AND (value < :2 OR :2 IS NULL)` so that you only bind(1) and bind(2) below

Reply:
Oooooh!